compiler/src/main/kotlin/io/github/apl_cornell/viaduct/syntax/circuit/Statements.kt

```diff
@@ -0,0 +1,196 @@
+import kotlinx.collections.immutable.PersistentList
+import kotlinx.collections.immutable.toPersistentList
+
+/** A computation with side effects. */
+sealed class StatementNode : Node()
+
+/**
+ * A statement that is _not_ a combination of other statements, and that does not affect
+ * control flow.
+ */
+sealed class SimpleStatementNode : StatementNode()
+
+sealed class CircuitStatementNode : SimpleStatementNode()
+
+class IndexParameterNode(
+    override val name: VariableNode,
+    val bound: IndexExpressionNode,
+    override val sourceLocation: SourceLocation
+) : Node(), VariableDeclarationNode {
+    override fun toDocument(): Document = name.toDocument() * "<" * bound
+}
+
+// Blocks
+
+/* Non-circuit statement
+/** A sequence of statements. */
+sealed class BlockNode(
+    open val statements: PersistentList<StatementNode>,
+    override val sourceLocation: SourceLocation
+) : StatementNode(), List<StatementNode> by statements {
+    override fun toDocument(): Document {
+        val statements: List<Document> = statements.map {
+            if (it is SimpleStatementNode)
+                it.toDocument()
+            else
+                it.toDocument()
+        }
+        val body: Document = statements.concatenated(separator = Document.forcedLineBreak)
+        return Document("{") +
+            (Document.forcedLineBreak + body).nested() + Document.forcedLineBreak + "}"
+    }
+}
+ */
+
+/** A sequence of circuit statements. */
+class CircuitBlockNode
+private constructor(
+    val statements: PersistentList<CircuitStatementNode>,
+    val returnStatement: ReturnNode,
+    override val sourceLocation: SourceLocation
+) : CircuitStatementNode(), List<CircuitStatementNode> by statements {
+    constructor(statements: List<CircuitStatementNode>, returnStatement: ReturnNode, sourceLocation: SourceLocation) :
+        this(statements.toPersistentList(), returnStatement, sourceLocation)
+
+    override fun toDocument(): Document {
+        val statements: MutableList<Document> = (statements.map { it.toDocument() } as MutableList<Document>)
+        statements.add(returnStatement.toDocument())
+        val body: Document = statements.concatenated(separator = Document.forcedLineBreak)
+        return Document("{") +
+            (Document.forcedLineBreak + body).nested() + Document.forcedLineBreak + "}"
+    }
+}
+
+// Simple Statements
+
+/** Binding the result of an expression to a variable. */
+class LetNode(
+    override val name: VariableNode,
+    val indices: Arguments<IndexParameterNode>,
+    val value: ExpressionNode,
+    override val sourceLocation: SourceLocation
+) : CircuitStatementNode(), VariableDeclarationNode {
+    override fun toDocument(): Document =
+        keyword("val") * name + indices.bracketed() * "=" * value
+}
+
+class ReturnNode(
+    val values: Arguments<PureExpressionNode>,
+    override val sourceLocation: SourceLocation
+) : CircuitStatementNode() {
```

Comment:
This should not be a circuit statement. We don't want these appearing in the middle of blocks.